Add documentation for migrating systems for the 2023-07 release (#148)
ni/install-systemlink-enterprise: release-notes/2023-07/SystemsConnectionMigration.md

# **Migration plan for systems connectivity that is affected by the breaking change in the salt-master chart.**

## The current release is introducing a breaking change regarding the systems connectivity. This means that any SLE <=0.14 release is affected by this change. The saltmaster service which manages the connectivity with the client systems has been replaced with a new implementation which causes all the connected clients to become disconnected. The next steps will describe how to re-approve and re-connect all the previously connected clients

&nbsp;

## **Before doing the upgrade of SLE**

The old salt-master has a pair of RSA keys that are used for securing the communication. The public key is already cached on all the
accepted clients. We need to persist these keys between all the SLE versions, otherwise the client systems will reject the new salt-master
that would have a different public key.

> 1. Open a terminal to the current "saltmaster-0" pod.

Reviewer: Generally, ordered lists in markdown can use `1.` for all items. This makes it lower churn if you need to reorder or insert new items in the list. e.g. 

```
1. foo
1. bar
1. bam
```

Is rendered as:
1. foo
1. bar
1. bam

2. Navigate to /etc/salt/pki/master

> 3. Grab the master.pem (private key) and master.pub (public key) values.

Author: No, these are required inputs to the helm chart, in a similar way we define the mongo credentials. Even if it is adding one additional manual step for generating these keys, I think the helm chart definition is cleaner this way. An alternative would have been an init-container that would check whether it needs to create the secret or not. Are you concerned about this?

4. Set these values to the secrets definition object of the saltmaster chart in the top level secrets values yaml file. Make sure
you use the "|" indicator because these are multiline strings.

> The new helm chart will create a kubernetes secret using these 2 keys and the pods will use it.

Reviewer: ```suggestion
The updated chart will create a Kubernetes secret using these two keys.
```

> &nbsp;

Reviewer: This can be removed

> ## **After doing the upgrade of SLE**

Reviewer: ```suggestion
## **Post SLE Upgrade**
```

> The expectation is that all the systems that prior the upgrade were connected (green/orange) are now disconnected and in the

Reviewer: ```suggestion
After the upgrade, it is expected that all previously connected connected (green/orange systems will be shown as **Pending**.
```

Author: The actual state will be "Disconnected" and this is how they will be shown in the grid. But in addition to that, all those clients will also be in the Pending systems list. I do really want to point out that they will be disconnected because otherwise, we will not be able to reconnect them properly.

Maybe this is better?
After the upgrade, it is expected that all previously connected (green/orange systems) will be shown as **Disconnected**  . In addition to that, all these clients will also be shown in the **Pending systems** list.

> The expectation is that all the systems that prior the upgrade were connected (green/orange) are now disconnected and in the

Reviewer: We likely need to reference the permissions needed by the user who is going to run the notebook for that operation to be successful.

> pending list. The pending list is only displaying us minion ids which are hard to identify. For re-approving all those systems we will use a Jupyter Notebook to automate the operation. All the pending clients that are currently managed will be re-approved in their current workspaces.

Reviewer: > The pending list is only displaying us minion ids which are hard to identify.

I don't understand what we mean by this line. Do we really need it?

> pending list. The pending list is only displaying us minion ids which are hard to identify. For re-approving all those systems we will use a Jupyter Notebook to automate the operation. All the pending clients that are currently managed will be re-approved in their current workspaces.

Reviewer: ```suggestion
A [Jupyter Notebook](include-direct-link) has been provided if you wish to automate system approval. This notebook will add systems to the workspace they were previously a member of.
```

> 1. Make sure all clients are in the disconnected state and they appear in the pending systems list.

Reviewer: ```suggestion
1. Ensure all systems are in the disconnected state and appear in the pending systems list.
```

> 2. Upload the "NewSaltMaster-SystemsMigration.ipynb" notebook to JupyterHub.

Reviewer: ```suggestion
2. Upload **NewSaltMaster-SystemsMigration.ipynb** notebook to JupyterHub.
```

> 3. Find a "guinea pig" system and make sure this system was connected and now is in the pending systems list.

Reviewer: ```suggestion
3. Identify a single system to validate this operation of this notebook..
```

> 4. Override "managed_systems_to_approve" with it's id and workspace. The notebook already

Reviewer: ```suggestion
4. Override `managed_systems_to_approve` with it's id and workspace. The notebook already
```

> 4. Override "managed_systems_to_approve" with it's id and workspace. The notebook already

Reviewer: This also feels like a good point to show a code snippet with the change we anticipate the user to make.

contains this line, we only need to uncomment it and add the id and workspace.

> 5. Run the notebook - check the output, it should say that 1 system was approved.

Reviewer: We can likely reference the exact string output we expect in this line.

> 6. The client should become connected (green). One thing that could prevent that is the improper

Reviewer: ```suggestion
6. Validate that the selected system is now connected (green). 
    **Note** Misconfiguration of the RSA keys for the salt master can lead to this operation failing. Review the new secrets and [logs on the system](https://knowledge.ni.com/KnowledgeArticleDetails?id=kA00Z000000kGcSSAU&l=en-US)

One thing that could prevent that is the improper
```

configurationrsa of the rsa keys of the saltmaster. If it doesn't become connected we need to check the secret and
the logs from the client side to better understand why it is not gettign connected.

> 7. After connecting the "guinea pig" system we can now uncomment the override of "managed_systems_to_approve" and run the notebook again.

Reviewer: ```suggestion
7. After connecting this single system, you can now uncomment the override of `managed_systems_to_approve` and run the notebook again.
```

> 7. After connecting the "guinea pig" system we can now uncomment the override of "managed_systems_to_approve" and run the notebook again.

Reviewer: This is another opportunity to add a code snippet that shows exactly what we intend on the user to change.

> 8. The new run will re-approve all the systems in batches of 50 (this can also be modified in the notebook)

Reviewer: This feels more like a note than a step. The final steps should likely be validating the operation was successful.

# Migrating Connected Systems to SLE 2023-07 or later

> The current release is introducing a breaking change regarding the systems connectivity. This means that SLE releases prior to SLE 2023-07 are affected by this change. The saltmaster service which manages the connectivity with the client systems has been replaced with a new implementation which causes all the connected clients to become disconnected. The next steps will describe how to re-approve and re-connect all the previously connected clients

Reviewer 2: ```suggestion
The current release is introduces a breaking change regarding the systems connectivity that affects releases prior to SLE 2023-07. The saltmaster service which manages the connectivity with the client systems has been replaced with a new implementation that disconnects all connected clients. Complete the following steps to re-approve and re-connect all previously connected clients.
```

## Prior to upgrading to SLE 2023-07

> The old salt-master has a pair of RSA keys that are used for securing the communication. The public key is already cached on all the

Reviewer 2: ```suggestion
The old salt-master uses a pair of RSA keys to secure communication. The public key is already cached on all
```

> accepted clients. We need to persist these keys between all the SLE versions, otherwise the client systems will reject the new salt-master

Reviewer 2: ```suggestion
accepted clients. You must persist these keys between all versions of SLE, so the client systems do not reject the new salt-master
```

that would have a different public key.

1. Open a terminal to the current 'saltmaster-0' pod.
2. Navigate to '/etc/salt/pki/master'
3. Grab the master.pem (private key) and master.pub (public key) values.
4. Set these values to the secrets definition object of the saltmaster chart in the top level secrets values yaml file. Make sure
you use the '|' indicator because these are multiline strings. See more info here: <https://yaml-multiline.info/>

The updated chart will create a Kubernetes secret using these two keys.

## Post SLE Upgrade

After the upgrade, it is expected that all previously connected (green/orange systems) will be shown as **Disconnected**. In addition to that, all these clients will also be shown in the **Pending systems** list. A [Jupyter Notebook](./NewSaltMaster-SystemsMigration.ipynb) has been provided if you wish to automate system approval. This notebook will add systems to the workspace they were previously a member of. The user that will run the notebook needs to have the **Add systems** privilege in all the necessary workspaces.  

> 1. Ensure all systems are in the disconnected state and they appear in the pending systems list.

Reviewer 2: ```suggestion
1. Ensure all systems are disconnected and appear in the pending systems list.
```

> 1. Upload **NewSaltMaster-SystemsMigration.ipynb** notebook to JupyterHub.

Reviewer 2: ```suggestion
1. Upload **NewSaltMaster-SystemsMigration.ipynb** to JupyterHub.
```

> 1. Identify a single system to validate this operation of this notebook.

Reviewer 2: ```suggestion
1. Identify a single system to validate the notebook operation.
```

> 1. Override `managed_systems_to_approve` with it's id and workspace. The notebook already

Reviewer 2: ```suggestion
1. Override `managed_systems_to_approve` with its id and workspace. The notebook already
```

contains this line, we only need to uncomment it and add the id and workspace:

    ```python
    managed_systems_to_approve = [{"id":"Precision_5570--SN-GPP8WT3--MAC-15-75-5B-DC-5F-BB","workspace":"846e294a-a007-47ac-9fc2-fac07eab240a"}]
    ```

> 1. Run the notebook - check the output, it should say that 1 system was approved.

Reviewer 2: ```suggestion
1. Run the notebook and check the output. It should say that one system was approved.
```

    ```python
    Approving 1 batches of 50 systems
    Approving 1 minions, batch: 1
    [{'action': 'ACCEPT',
    'id': 'Precision_5570--SN-GPP8WT3--MAC-15-75-5B-DC-5F-BB',
    'key': None,
    'workspace': '846e294a-a007-47ac-9fc2-fac07eab240a'}]
    Manage Systems Keys Result: {} {'error': None}
    ```

1. Validate that the selected system is now connected (green).

>     **Note** Misconfiguration of the RSA keys for the salt master can lead to this operation failing. Review the new secrets and [logs on the system](https://knowledge.ni.com/KnowledgeArticleDetails?id=kA00Z000000kGcSSAU&l=en-US)

Reviewer 2: ```suggestion
    **Note** Misconfiguration of the RSA keys for the salt master may cause this operation to fail. Review the new secrets and [logs on the system](https://knowledge.ni.com/KnowledgeArticleDetails?id=kA00Z000000kGcSSAU&l=en-US)
```
